kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/PartitionUtils.java

```diff
@@ -51,76 +51,60 @@ public class PartitionUtils {
-   *     physical schema.
-   * @return A new {@link StructType} representing the physical schema without the specified
-   *     partition columns.
+   * @param dataBatch Data batch to which the partition columns will be added.
+   * @param logicalSchema Logical schema of the table scan. Used to insert partition columns at the
+   *     right positions in the data batch. This logical schema must contain column mapping metadata
+   *     if column mapping is enabled.
+   * @param partitionValues Map of partition column name to value.
+   * @param expressionHandler Expression handler used to evaluate the partition values.
+   * @return A new {@link ColumnarBatch} with the partition columns added.
    */
-  public static StructType physicalSchemaWithoutPartitionColumns(
-      StructType logicalSchema, StructType physicalSchema, Set<String> columnsToRemove) {
-    if (columnsToRemove == null || columnsToRemove.isEmpty()) {
-      return physicalSchema;
-    }
-
-    // Partition columns are top-level only
-    Map<String, String> physicalToLogical =
-        new HashMap<String, String>() {
-          {
-            IntStream.range(0, logicalSchema.length())
-                .mapToObj(i -> new Tuple2<>(logicalSchema.at(i), physicalSchema.at(i)))
-                .forEach(tuple2 -> put(tuple2._2.getName(), tuple2._1.getName()));
-          }
-        };
-
-    return new StructType(
-        physicalSchema.fields().stream()
-            .filter(field -> !columnsToRemove.contains(physicalToLogical.get(field.getName())))
-            .collect(Collectors.toList()));
-  }
-
   public static ColumnarBatch withPartitionColumns(
-      ExpressionHandler expressionHandler,
       ColumnarBatch dataBatch,
+      StructType logicalSchema,
       Map<String, String> partitionValues,
-      StructType schemaWithPartitionCols) {
-    if (partitionValues == null || partitionValues.size() == 0) {
-      // no partition column vectors to attach to.
+      ExpressionHandler expressionHandler) {
+    if (partitionValues == null || partitionValues.isEmpty()) {
+      // No partition column vectors to attach to the data batch
       return dataBatch;
     }
 
-    for (int colIdx = 0; colIdx < schemaWithPartitionCols.length(); colIdx++) {
-      StructField structField = schemaWithPartitionCols.at(colIdx);
-
-      if (partitionValues.containsKey(structField.getName())) {
-        // Create a partition vector
-
-        ColumnarBatch finalDataBatch = dataBatch;
+    int numPartitionColumnsInSchema =
+        (int)
+            logicalSchema.fields().stream()
+                .map(ColumnMapping::getPhysicalName)
+                .filter(partitionValues::containsKey)
+                .count();
+    if (logicalSchema.length() - numPartitionColumnsInSchema != dataBatch.getSchema().length()) {
+      throw DeltaErrors.logicalPhysicalSchemaMismatch(
+          logicalSchema.length(), partitionValues.size(), dataBatch.getSchema().length());
+    }
```

Comment:
I think we can simplify this check to just do
`partitionValues.size + dataBatch.getSchema().length() == logicalSchema.length()`

Reply:
if you wanted to you could check that they keySets match for the above formula but I think that might be overkill and we can rely/assume that this was used correctly

Reply:
I had [a similar discussion](https://github.com/delta-io/delta/pull/5007#discussion_r2251230654) with @johanl-db and we agreed that the check you mentioned does not need to hold because `partitionValues` contains all partition columns of a table but the user/engine does not necessarily request all partition columns in a scan.